Feat/add linenoise library

SoucheSouche opened the pull request:
# Checklist

- [x] Component contains License
- [x] Component contains README.md
- [x] Component contains idf_component.yml file with `url` field defined
- [x] Component was added to [upload job](https://github.com/espressif/idf-extra-components/blob/master/.github/workflows/upload_component.yml#L18)
- [x] _Optional:_ Component contains unit tests
- [ ] CI passing

# Change description
Add linenoise to the list of extra components. The library is used in the console component of esp-idf that is going through refactoring. One part of the refactoring is to move third party libraries outside of the console component. Eventually, the console component itself will be removed from esp-idf, reason why this PR adds the linenoise lib to the idf-extra-components.

suda-morris commented:
There was a PR about the linenoise library: https://github.com/espressif/esp-idf/pull/10526
WDYT? @SoucheSouche

SoucheSouche commented:
> There was a PR about the linenoise library: [espressif/esp-idf#10526](https://github.com/espressif/esp-idf/pull/10526) WDYT? @SoucheSouche

Yes, I am aware of it. I took into account Ivan's comment left in the PR when implementing this version of linenoise. The author of the PR you mentioned is aware that I will take over and continue the work on improving the console component (linenoise and more generally the list of idea he listed here https://github.com/espressif/esp-idf/issues/9986)

igrr commented on linenoise/linenoise/linenoise.h:
```diff
@@ -0,0 +1,335 @@
+    linenoise_read_bytes_fn read_bytes_fn;
+    linenoise_write_bytes_fn write_bytes_fn;
+    char **history;
+} linenoise_instance_param_t;
```
For consistency with other components, let's call this "config"

```suggestion
} linenoise_instance_config_t;
```

igrr commented on linenoise/linenoise/linenoise.h:
```diff
@@ -0,0 +1,335 @@
+ * instance
+ */
+typedef struct linenoise_instance_param {
+    const char *prompt;
```
Please add Doxygen comments for struct members.

igrr commented on linenoise/linenoise/linenoise.h:
```diff
@@ -0,0 +1,335 @@
+ * @brief Structure defining the parameters needed by a linenoise
+ * instance
+ */
+typedef struct linenoise_instance_param {
```
Instance config structure should have a `void* user_ctx` member that will be passed to all the callbacks.

This design in common in libraries which support multiple instances, and allows, for example, for the caller to find the local "state" or "context" structure related to the instance which has triggered the particular callback. This is also useful when wrapping the library in a C++ class, the `user_ctx` is then often the pointer to `this` (class instance).

SoucheSouche replied:
Implemented.

igrr commented on linenoise/linenoise/linenoise.h:
```diff
@@ -0,0 +1,335 @@
+ *
+ * @return A populated linenoise_instance_param_t structure with default values.
+ */
+linenoise_instance_param_t linenoise_get_instance_param_default(void);
```
```suggestion
void linenoise_get_instance_param_default(linenoise_instance_param_t *out);
```

The caller may have placed the configuration structure in heap or .bss, allowing them to pass the pointer to it helps avoid storing another copy on the stack.

igrr commented on linenoise/linenoise/linenoise.h:
```diff
@@ -0,0 +1,335 @@
+ * @param param Pointer to the configuration parameters for the instance.
+ * @return Handle to the created instance, or NULL on failure.
+ */
+linenoise_handle_t linenoise_create_instance(linenoise_instance_param_t *param);
```
```suggestion
esp_err_t linenoise_create_instance(const linenoise_instance_param_t *param, linenoise_handle_t* out_handle);
```

1. Configuration structure can be a const pointer
2. You may need to return different types of errors (out of RAM, invalid argument in the config structure, etc), which simply returning NULL doesn't allow. Hence, return esp_err_t and get the instance handle as an output argument.

igrr commented on linenoise/linenoise/linenoise.h:
```diff
@@ -0,0 +1,335 @@
+ * @param handle Handle to the instance to delete.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_delete_instance(linenoise_handle_t handle);
```
What should the caller do if the "delete instance" function has returned an error?

igrr commented on linenoise/linenoise/linenoise.h:
```diff
@@ -0,0 +1,335 @@
+ * @param empty_line true to allow empty lines, false to disallow.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_set_empty_line(linenoise_handle_t handle, bool empty_line);
```
Is there a use case for changing these settings (empty_line, multi_line, dumb_mode, max_cmd_line_length) once the instance is already created? If there is no concrete use case, it might be simpler to not implement these getters and setters.

SoucheSouche replied:
If the user creates a linenoise instance and uses it for a given console instance, then if the user wants to delete the console instance used with the linenoise instance and create a new console instance with different linenoise specs, it could be worth keeping those getters and setters. Although if we follow through with this idea we might want to add other getters and setters to be able to change other config fields (e.g., prompt).

Although deleting and creating a new line oise instance will do the job.

igrr commented on linenoise/linenoise/linenoise.h:
```diff
@@ -0,0 +1,335 @@
+ *OLD LINENOISE API
+ *------------------------------------------------------------------------------------------------*/
+
+typedef struct linenoise_completions linenoiseCompletions;
```
I'd recommend having the compatibility layer in a different header file from the "new" API.

Components which use the old API, can include, for example, "linenoise/linenoise.h" (as they used to).

Components which use the new API can include a different file, such as "esp_linenoise.h", and only get the new functions, without the old functions polluting the global namespace.

SoucheSouche replied:
topology updated. I added the new linenoise to esp_linenoise.c/h files

igrr commented on linenoise/idf_component.yml:
```diff
@@ -0,0 +1,9 @@
+version: "1.0.0"
```
Since the original upstream version of linenoise was tagged as 1.0, perhaps we can define this as 2.0.0?

igrr commented on linenoise/sbom_linenoise.yml:
```diff
@@ -0,0 +1,3 @@
+name: sbom_linenoise
+description: Line editing C library
+url: https://github.com/antirez/linenoise
```
The library is not maintained upstream anymore, i think the URL should point to idf-extra-components instead.

Please also check the other fields:

```yml
name: linenoise
version: 2.0.0
cpe: cpe:2.3:a:espressif:linenoise:{}:*:*:*:*:*:*:*
supplier: 'Organization: Espressif Systems'
```

SoucheSouche commented:
@igrr, I updated the component based on your input. I left a few comments open for discussion.

I added a bunch of tests for the basic functionality of linenoise when escape sequences are supported by the terminal.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/idf-extra-components?pullRequest=526) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/espressif/idf-extra-components?pullRequest=526) before we can accept your contribution.<br/>**4** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: Ashish285<br/>:white_check_mark: suda-morris<br/>:white_check_mark: Kainarx<br/>:white_check_mark: SoucheSouche<br/>:x: wanckl<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/idf-extra-components?pullRequest=526) it.</sub>

igrr commented on esp_linenoise/include/esp_linenoise.h:
```diff
@@ -0,0 +1,303 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
```
Please convert the above license block to SPDX format. And you can relicense the final library as Apache 2.0.

igrr commented on esp_linenoise/include/esp_linenoise.h:
```diff
@@ -0,0 +1,303 @@
+    esp_linenoise_free_hints_callback free_hints_cb; /*!< Callback function to free hints returned by `hints_cb` */
+    void *user_ctx; /*!< User context information passed to the callbacks */
+    esp_linenoise_read_bytes_fn read_bytes_fn; /*!< Function used to read bytes from the input stream */
+    esp_linenoise_write_bytes_fn write_bytes_fn; /*!< Function used to write bytes to the output stream */
```
Why are some callback names end with `_callback` and others with `_fn`? Can we make them consistent?

I would also recommend keeping `_t` suffix for function pointer typedefs.

igrr commented on esp_linenoise/include/esp_linenoise.h:
```diff
@@ -0,0 +1,303 @@
+/**
+ * @brief Returns the default parameters for creating a linenoise instance.
+ *
+ * @param config esp_linenoise_config_t structure to populate with default values.
```
```suggestion
 * @param[out] config esp_linenoise_config_t structure to populate with default values.
```

igrr commented on esp_linenoise/include/esp_linenoise.h:
```diff
@@ -0,0 +1,303 @@
+ * @param handle Handle to the linenoise instance.
+ * @param cmd_line_buffer Buffer to store the input line.
+ * @param cmd_line_length Length of the command line buffer.
+ * @return ESP_OK on success, or error code on failure.
```
It would be nice to clarify what happens if cmd_line_buffer is not large enough to fit a line of input. Do we return ESP_OK or an error, is the line truncated, etc.

SoucheSouche commented on esp_linenoise/include/esp_linenoise.h:
```diff
@@ -0,0 +1,276 @@
+    const char *prompt; /*!< Prompt string displayed to the user */
+    size_t max_cmd_line_length; /*!< Maximum length (in bytes) of the input command line */
+    int history_max_length; /*!< Maximum number of entries to store in command history */
+    int in_fd; /*!< File descriptor to read input from (e.g., STDIN_FILENO) */
```
I think I will do it as follow:
- if the user doesn't provide a custom read, then linenoise will set the read mode to blocking internally. If the user provides its own custom read function, then it will fall under the user responsibility to set the proper read mode,

what do you think ?

SoucheSouche commented:
@igrr, could you please take another look, and check the comments I left opened?
